### PR TITLE
Ethereum: Add function signatures to generated contract call code

### PR DESCRIPTION
**Pull request by Jannis**
Part of resolving https://github.com/graphprotocol/support/issues/48.

In order to disambiguate which overloaded version of a function is being called, the function signature is passed to the host along with the call. Graph Node can then match it to the correct function definition in the ABI.

### CODE REVIEW

File: src/abi.js [hunk 1]
```diff
@@ -34,17 +34,31 @@ module.exports = class ABI {
...
+      .get('inputs', [])
       .map(input => buildOldSignatureParameter(input))
       .join(',')})`
   }
 
   static eventSignature(event) {
-    return `${event.get('name')}(${(event.get('inputs') || [])
+    return `${event.get('name')}(${event
+      .get('inputs', [])
       .map(input => buildSignatureParameter(input))
```
Comment (fordN) [inline src/abi.js]: and here?

File: src/abi.js [hunk 2]
```diff
@@ -34,17 +34,31 @@ module.exports = class ABI {
   }
 
   static oldEventSignature(event) {
-    return `${event.get('name')}(${(event.get('inputs') || [])
+    return `${event.get('name')}(${event
+      .get('inputs', [])
       .map(input => buildOldSignatureParameter(input))
```
Comment (fordN) [inline src/abi.js]: Can the simplified notation used in `functionSignature()` be used here as well? 
```
.map(buildSignatureParameter)
```

File: src/abi.js [hunk 3]
```diff
@@ -34,17 +34,31 @@ module.exports = class ABI {
...
+  functionSignature(fn) {
+    let inputs = fn
+      .get('inputs', [])
+      .map(buildSignatureParameter)
+      .join(',')
+    let outputs = fn
+      .get('outputs', [])
+      .map(buildSignatureParameter)
+      .join(',')
+    return `${fn.get('name')}(${inputs})${outputs.length > 0 ? `:(${outputs})` : ''}`
```
Comment (fordN) [inline src/abi.js]: It could be helpful for readability to include a comment explaining the function signature format since the `return ...` line is a bit hard to parse.

**Review by fordN — COMMENTED** [review]
After reviewing I only have a few minor comments on readability and consistency. I also left a reminder to update the dependency reference for `graph-ts` once [PR#100](https://github.com/graphprotocol/graph-ts/pull/100#pullrequestreview-342861075) is merged.

**Comment by fordN** [comment]
The tests are not passing for me locally on both this branch and master though they _were_ successful during the travis build.
The `Init > From example` test is failing:

```
Expected: "
    Subgraph user/example-subgraph created in from-example·
    Next steps:·
      1. Run `graph auth https://api.thegraph.com/deploy/ <access-token>`
         to authenticate with the hosted service. You can get the access token from
         https://thegraph.com/explorer/dashboard/.·
      2. Type `cd from-example` to enter the subgraph.·
      3. Run `yarn deploy` to deploy the subgraph to
         https://thegraph.com/explorer/subgraph/user/example-subgraph.·
    Make sure to visit the documentation on https://thegraph.com/docs/ for further information.
    "
    Received: ""
```

File: src/abi.js (hunk 1 above)
Comment (Jannis) [reply src/abi.js]: Yup, done

File: src/abi.js (hunk 2 above)
Comment (Jannis) [reply src/abi.js]: Yes, done

File: src/abi.js (hunk 3 above)
Comment (Jannis) [reply src/abi.js]: I've added the following comment above the method:
```js
  /**
   * For the ABI of a function, returns a string function signature compatible
   * with the Rust `ethabi` library. It is of the form
   *
   *     <function>([<input-type-1>, ...])[:(<output-type-1,...)]
   *
   * A few examples for a function called `example`:
   *
   * - No inputs or outputs: `example()`
   * - One input and output: `example(uint256):(bool)`
   * - Multiple inputs and outputs: `example(uint256,(string,bytes32)):(bool,uint256)`
   */
```

**Comment by Jannis** [comment]
@fordN We should figure out why they don't run for you, but I just ran the test suite locally (`yarn test`) and it worked fine. So this is most likely unrelated to the PR.

File: src/abi.js (hunk 3 above)
Comment (fordN) [reply src/abi.js]: Perfect
